taf/testlib/virtual_env.py

```diff
@@ -353,33 +367,41 @@ def _delete_external_elements(self, routers_client=None, networks_client=None,
-            networks_client.delete_network(network_id)
+        net_2_router_map = self._get_external_elements(routers_client=routers_client, tenant_id=tenant_id)
+        for network_id, router_id in net_2_router_map.items():
+            if router_id and tenant_id and tenant_id == routers_client.show_router(router_id)['router']['tenant_id']:
+                self.class_logger.debug('Removing external router: (%s)',
+                                        routers_client.show_router(router_id)['router']['name'])
+                self.delete_router(router_id, routers_client, ports_client)
+
+            if tenant_id and tenant_id == networks_client.show_network(network_id)['network']['tenant_id']:
+                self.class_logger.debug('Removing external network: (%s)',
+                                        networks_client.show_network(network_id)['network']['name'])
+                networks_client.delete_network(network_id)
+            else:
+                self.class_logger.debug('Removing of external network: (%s) is not possible - try to reuse it',
+                                        networks_client.show_network(network_id)['network']['name'])
+                return network_id
+        return None
 
-    def _reuse_public_access(self, mgmt_net, routers_client=None):
+    def _reuse_public_access(self, external_net, routers_client=None, tenant_id=None):
         """
         @brief   Search for the external routers & networks
         """
         if not routers_client:
             routers_client = self.handle.os_adm.routers_client
 
         _net_cfg = self.config.network
-        router_2_net_map = self._get_external_elements(routers_client=routers_client)
-        if router_2_net_map:
-            for router_id, net_id in router_2_net_map.items():
-                rt_obj = routers_client.show_router(router_id)['router']
-                ext_ips = rt_obj['external_gateway_info']['external_fixed_ips']
-                for ip in ext_ips:
-                    if ip['ip_address'] in mgmt_net:
-                        _net_cfg.public_router_id, _net_cfg.public_network_id = router_id, net_id
-                        return True
+        net_2_router_map = self._get_external_elements(routers_client=routers_client, tenant_id=tenant_id)
+        if net_2_router_map:
+            for net_id, router_id in net_2_router_map.items():
+                if router_id and tenant_id and tenant_id == routers_client.show_router(router_id)['router']['tenant_id']:
+                    rt_obj = routers_client.show_router(router_id)['router']
+                    ext_ips = rt_obj['external_gateway_info']['external_fixed_ips']
+                    for ip in ext_ips:
+                        if ip['ip_address'] in external_net:
+                            _net_cfg.public_router_id, _net_cfg.public_network_id = router_id, net_id
+                            return True
```

Comment:
Please `continue` when possible, instead of nesting conditionals.

Please avoid repeating exactly similar function calls on successive lines.

```
for net_id, router_id in net_2_router_map.items():
    if not router_id or not tenant_id:
        continue
    rt_obj = routers_client.show_router(router_id)['router']
    if tenant_id != rt_obj['tenant_id']:
        continue
    ext_ips = rt_obj['external_gateway_info']['external_fixed_ips']
    with suppress(StopIteration):
        _net_cfg.public_router_id, _net_cfg.public_network_id = next((router_id, net_id) for ip in ext_ips if ip['ip_address'] in external_net)
        return True
```